File: pipeline/reimpl/corpt00c.py
"""
Reimplementation of CORPT00C — CardDemo Report Submission Screen.

CICS online program. Allows users to submit a transaction report job
(monthly, yearly, or custom date range). The COBOL original submits
JCL to an internal reader TDQ; this reimplementation captures the
parameters and returns a JobRequest descriptor that the caller can
use to invoke cbtrn03c or a similar batch report function.

Report types:
  Monthly → current month start → end of month
  Yearly  → current year Jan 1 → Dec 31
  Custom  → user-supplied start/end date (YYYY-MM-DD)

Navigation:
  ENTER → validate and submit job
  PF3   → return to COMEN01C
"""

from __future__ import annotations
import calendar
from dataclasses import dataclass
from datetime import date
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3,
)
from .csutldtc import validate_date


WS_PGMNAME  = "CORPT00C"
WS_TRANID   = "CR00"
DATE_FORMAT = "YYYY-MM-DD"


@dataclass
class JobRequest:
    report_name: str = ""
    start_date: str = ""
    end_date: str = ""


@dataclass
class ReportResult:
    job_request: Optional[JobRequest] = None
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    xctl_program: str = ""
    return_to_prev: bool = False


def process_report_screen(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    report_type: str,
    start_mm: str = "",
    start_dd: str = "",
    start_yyyy: str = "",
    end_mm: str = "",
    end_dd: str = "",
    end_yyyy: str = "",
    today: Optional[date] = None,
) -> ReportResult:
    """Process report submission screen — mirrors CORPT00C PROCEDURE DIVISION."""
    result = ReportResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea
    if today is None:
        today = date.today()

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result

    if eibaid in (DFHPF3, "PF3"):
        result.xctl_program = "COMEN01C"
        result.return_to_prev = True
        result.commarea = commarea
        return result

    if eibaid not in (DFHENTER, "ENTER"):
        result.error = True
        result.message = CCDA_MSG_INVALID_KEY
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    rt = report_type.strip().lower() if report_type else ""

    if rt == "monthly":
        yr = today.year
        mo = today.month
        start = f"{yr:04d}-{mo:02d}-01"
        last_day = calendar.monthrange(yr, mo)[1]
        end = f"{yr:04d}-{mo:02d}-{last_day:02d}"
        return _submit_job("Monthly", start, end, commarea, result)

    elif rt == "yearly":
        yr = today.year
        start = f"{yr:04d}-01-01"
        end   = f"{yr:04d}-12-31"
        return _submit_job("Yearly", start, end, commarea, result)

    elif rt == "custom":
        errors = []
        if not start_mm or not start_mm.strip():
            errors.append("Start Date - Month can NOT be empty...")
        if not start_dd or not start_dd.strip():
            errors.append("Start Date - Day can NOT be empty...")
        if not start_yyyy or not start_yyyy.strip():
            errors.append("Start Date - Year can NOT be empty...")
        if not end_mm or not end_mm.strip():
            errors.append("End Date - Month can NOT be empty...")
        if not end_dd or not end_dd.strip():
            errors.append("End Date - Day can NOT be empty...")
        if not end_yyyy or not end_yyyy.strip():
            errors.append("End Date - Year can NOT be empty...")

        if errors:
            result.error = True
            result.message = errors[0]
            result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
            result.screen.errmsg = result.message
            result.commarea = commarea
            return result

        start = f"{start_yyyy.strip()}-{start_mm.strip().zfill(2)}-{start_dd.strip().zfill(2)}"
        end   = f"{end_yyyy.strip()}-{end_mm.strip().zfill(2)}-{end_dd.strip().zfill(2)}"

        start_check = validate_date(start, DATE_FORMAT)
        if start_check.severity != 0:
            result.error = True
            result.message = "Start Date - Not a valid date..."
            result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
            result.screen.errmsg = result.message
            result.commarea = commarea
            return result

        end_check = validate_date(end, DATE_FORMAT)
        if end_check.severity != 0:
            result.error = True
            result.message = "End Date - Not a valid date..."
            result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
            result.screen.errmsg = result.message
            result.commarea = commarea
            return result

        return _submit_job("Custom", start, end, commarea, result)

    else:
        result.error = True
        result.message = "Select a report type to print report..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result


def _submit_job(report_name, start_date, end_date, commarea, result):
    result.job_request = JobRequest(
        report_name=report_name,
        start_date=start_date,
        end_date=end_date,
    )
    result.message = f"{report_name} report job submitted"
    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Corpt00c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.time.LocalDate;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;

/**
 * Java reimplementation of CORPT00C — CardDemo Report Submission Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/corpt00c.py}. Python is the source of truth.
 *
 * <p>Structure: standalone {@link #processReportScreen} method + thin
 * {@link #runVector} adapter.
 *
 * <p>Uses a fixed "today" of 2025-06-15 to match the Python adapter's
 * deterministic output.
 */
public class Corpt00c implements ProgramRunner {

    private static final String CCDA_MSG_INVALID_KEY = "Invalid key pressed.                    ";

    // ── Inner data types ────────────────────────────────────────────────────

    static class JobRequest {
        String reportName = "";
        String startDate = "";
        String endDate = "";

        JobRequest(String reportName, String startDate, String endDate) {
            this.reportName = reportName;
            this.startDate = startDate;
            this.endDate = endDate;
        }
    }

    static class ReportResult {
        JobRequest jobRequest = null;
        String message = "";
        boolean error = false;
        String xctlProgram = "";
        boolean returnToPrev = false;
    }

    // ── Standalone business method ──────────────────────────────────────────

    /**
     * Process report submission screen — mirrors {@code process_report_screen} in
     * {@code pipeline/reimpl/corpt00c.py}.
     */
    ReportResult processReportScreen(
            int eibcalen, String eibaid, int pgmContext,
            String reportType,
            String startMm, String startDd, String startYyyy,
            String endMm, String endDd, String endYyyy,
            LocalDate today) {

        ReportResult result = new ReportResult();

        if (eibcalen == 0) {
            result.returnToPrev = true;
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (pgmContext == 0) {
            // First entry — show blank form
            return result;
        }

        if ("PF3".equals(eibaid)) {
            result.xctlProgram = "COMEN01C";
            result.returnToPrev = true;
            return result;
        }

        if (!"ENTER".equals(eibaid)) {
            result.error = true;
            result.message = CCDA_MSG_INVALID_KEY;
            return result;
        }

        String rt = reportType != null ? reportType.trim().toLowerCase() : "";

        if ("monthly".equals(rt)) {
            int yr = today.getYear();
            int mo = today.getMonthValue();
            String start = String.format("%04d-%02d-01", yr, mo);
            int lastDay = YearMonth.of(yr, mo).lengthOfMonth();
            String end = String.format("%04d-%02d-%02d", yr, mo, lastDay);
            return submitJob("Monthly", start, end, result);
        } else if ("yearly".equals(rt)) {
            int yr = today.getYear();
            String start = String.format("%04d-01-01", yr);
            String end = String.format("%04d-12-31", yr);
            return submitJob("Yearly", start, end, result);
        } else if ("custom".equals(rt)) {
            // Validate fields
            List<String> errors = new ArrayList<>();
            if (startMm == null || startMm.trim().isEmpty())
                errors.add("Start Date - Month can NOT be empty...");
            if (startDd == null || startDd.trim().isEmpty())
                errors.add("Start Date - Day can NOT be empty...");
            if (startYyyy == null || startYyyy.trim().isEmpty())
                errors.add("Start Date - Year can NOT be empty...");
            if (endMm == null || endMm.trim().isEmpty())
                errors.add("End Date - Month can NOT be empty...");
            if (endDd == null || endDd.trim().isEmpty())
                errors.add("End Date - Day can NOT be empty...");
            if (endYyyy == null || endYyyy.trim().isEmpty())
                errors.add("End Date - Year can NOT be empty...");

            if (!errors.isEmpty()) {
                result.error = true;
                result.message = errors.get(0);
                return result;
            }

            String start = startYyyy.trim() + "-"
                    + padLeft(startMm.trim(), 2) + "-"
                    + padLeft(startDd.trim(), 2);
            String end = endYyyy.trim() + "-"
                    + padLeft(endMm.trim(), 2) + "-"
                    + padLeft(endDd.trim(), 2);

            if (!isValidDate(start)) {
                result.error = true;
                result.message = "Start Date - Not a valid date...";
                return result;
            }
            if (!isValidDate(end)) {
                result.error = true;
                result.message = "End Date - Not a valid date...";
                return result;
            }

            return submitJob("Custom", start, end, result);
        } else {
            result.error = true;
            result.message = "Select a report type to print report...";
            return result;
        }
    }

    private ReportResult submitJob(String reportName, String startDate, String endDate, ReportResult result) {
        result.jobRequest = new JobRequest(reportName, startDate, endDate);
        result.message = reportName + " report job submitted";
        return result;
    }

    private static String padLeft(String s, int len) {
        while (s.length() < len) s = "0" + s;
        return s;
    }

    private static boolean isValidDate(String dateStr) {
        try {
            LocalDate.parse(dateStr, DateTimeFormatter.ISO_LOCAL_DATE);
            return true;
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    // ── runVector adapter ───────────────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "FIRST_ENTRY");

        LocalDate fixedToday = LocalDate.of(2025, 6, 15);

        ReportResult result;

        switch (scenario) {
            case "FIRST_ENTRY":
                result = processReportScreen(100, "ENTER", 0,
                        "", "", "", "", "", "", "", fixedToday);
                break;
            case "MONTHLY_REPORT":
                result = processReportScreen(100, "ENTER", 1,
                        "monthly", "", "", "", "", "", "", fixedToday);
                break;
            case "YEARLY_REPORT":
                result = processReportScreen(100, "ENTER", 1,
                        "yearly", "", "", "", "", "", "", fixedToday);
                break;
            case "CUSTOM_REPORT":
                result = processReportScreen(100, "ENTER", 1,
                        "custom", "01", "01", "2025", "06", "30", "2025", fixedToday);
                break;
            case "INVALID_TYPE":
                result = processReportScreen(100, "ENTER", 1,
                        "", "", "", "", "", "", "", fixedToday);
                break;
            case "PF3_RETURN":
                result = processReportScreen(100, "PF3", 1,
                        "", "", "", "", "", "", "", fixedToday);
                break;
            default:
                result = processReportScreen(100, "ENTER", 0,
                        "", "", "", "", "", "", "", fixedToday);
                break;
        }

        String jobName = "";
        String startDate = "";
        String endDate = "";
        if (result.jobRequest != null) {
            jobName = result.jobRequest.reportName;
            startDate = result.jobRequest.startDate;
            endDate = result.jobRequest.endDate;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram != null ? result.xctlProgram : "");
        out.put("RETURN_TO_PREV", result.returnToPrev ? "Y" : "N");
        out.put("JOB_NAME", jobName);
        out.put("START_DATE", startDate);
        out.put("END_DATE", endDate);
        return out;
    }
}
